markdown/blog/shai-hulud-postmortem.md

```diff
@@ -0,0 +1,101 @@
+
+## Incident Timeline
+
+ **Mon, 24 Nov 2025**
+  - `03:58:47 UTC` — Pushing of malicious package versions to npm registry.
+
+  - `10:02:00 UTC` — [Report by Charlie Eriksen](https://www.aikido.dev/blog/shai-hulud-strikes-again-hitting-zapier-ensdomains) from Aikido about compromised packages.
+  - `10:32:00 UTC` — Acknowledgment of the issue by AsyncAPI team, and communication initiated to inform the community.
+  - `10:42:00 UTC` — Initial investigation and revoking of NPM tokens.
+  - `10:58:00 UTC` — Reported the incident to npm security team and started deprecating affected packages as unpublishing is not allowed with dependent packages.
+  - `11:46:00 UTC` — Action taken by NPM security team to unpublish the malicious packages.
+  - `13:33:00 UTC` — Suspected leak found in [vs-asyncapi-preview](https://github.com/asyncapi/vs-asyncapi-preview), as malicious version (1.0.1) was pushed to [OpenVSX Registry](https://open-vsx.org/extension/asyncapi/asyncapi-preview) through report from [Yusuke Sugamiya](https://x.com/DNPP)
+  - `13:45:00 UTC` — Revoked OpenVSX token and reported to OpenVSX security team.
+  - `14:30:00 UTC` — OpenVSX team unpublished the malicious package version.
+  - `15:45:00 UTC` — Incident report writing started.
+
+## What Happened
+
+On the morning of November 24, 2025, we were alerted by Charlie Eriksen from Aikido about suspicious activity involving several AsyncAPI-related packages on the npm registry. Upon investigation, we discovered that unauthorized versions of our packages had been published, containing malicious code designed to exfiltrate sensitive information from users' environments. This was further verified by Charlie who identified it as a Shai-Hulud worm attack. 
+
+Simultaneously, we were notified by [Yusuke Sugamiya](https://x.com/DNPP) that one of our Visual Studio Code extensions, [vs-asyncapi-preview](https://open-vsx.org/extension/asyncapi/asyncapi-preview) had also been compromised, with a malicious version (1.0.1) being published to the OpenVSX registry.
+
+All AsyncAPI packages had malicious versions published with a pattern of two bad patch releases. Example: for package version 1.2.3, versions 1.2.4 and 1.2.5 were malicious. The malicious code was designed to collect environment variables and send them to an external server controlled by the attacker. More on the technical details can be found in [Aikido's detailed analysis](https://www.aikido.dev/blog/shai-hulud-strikes-again-hitting-zapier-ensdomains). 
+
+This led to a spread onto other dependent packages and services, including Postman, Zapier and ENS Domains. The worm works by scanning environment variables for sensitive information and publishing them in user's public repositories. The code for worm propagation is located in `bun_environment.js` file within the malicious package versions. And the same is run through `setup_bun.js` during package postinstall script execution.
```

Comment:
@Shurtu-gal I think this statement, `This led to a spread onto other dependent packages and services, including Postman, Zapier and ENS Domains.`  is misleading since AsyncAPI wasn't the propagation source to the other attacks but rather Patient 0 / first victim. The statement makes it seem that the worm got propagated from AsyncAPI being compromised. Maybe you can reword this?

Reply:
Sure.